src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -148,27 +176,46 @@ public static string HtmlDecode(string value)
                 return value;
             }
 
-            // Don't create StringBuilder if we don't have anything to encode
             if (!StringRequiresHtmlDecoding(value))
             {
                 return value;
             }
 
-            StringBuilder sb = StringBuilderCache.Acquire(value.Length);
-            HtmlDecode(value, sb);
-            return StringBuilderCache.GetStringAndRelease(sb);
+            var sb = new ValueStringBuilder(value.Length);
+            HtmlDecode(value, ref sb);
+            return sb.ToString();
         }
 
         public static void HtmlDecode(string value, TextWriter output)
         {
-            output.Write(HtmlDecode(value));
+            if (output == null)
+            {
+                throw new ArgumentNullException(nameof(output));
+            }
+            if (value == null)
+            {
+                throw new ArgumentNullException(nameof(value));
+            }
+            if (value.Length == 0)
+            {
+                return;
+            }
+
+            if (!StringRequiresHtmlDecoding(value))
+            {
+                output.Write(value);
+                return;
+            }
+
+            var sb = new ValueStringBuilder(value.Length);
+            HtmlDecode(value, ref sb);
+            output.Write(sb.AsReadOnlySpan());
+            sb.Dispose();
         }
 
         [SuppressMessage("Microsoft.Usage", "CA1806:DoNotIgnoreMethodResults", MessageId = "System.UInt16.TryParse(System.String,System.Globalization.NumberStyles,System.IFormatProvider,System.UInt16@)", Justification = "UInt16.TryParse guarantees that result is zero if the parse fails.")]
```

Comment:
Since you are optimizing allocations, you can also replace `uint.TryParse(value.Substring` below with `uint.TryParse(value.AsSpan` below; and delete this SuppressMessage - it does not seem to be relevant anymore.

Reply:
[Done](https://github.com/dotnet/corefx/pull/27250/commits/b1b17d2a0e58e863e86440651b188ac3aab86c77#diff-831332bbb389db07350457a4d92095f8L246), good spot, thanks!